common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java

```diff
@@ -158,6 +158,34 @@ public void pushBlocks(
     }
   }
 
+  @Override
+  public void finalizeShuffleMerge(
+      String host,
+      int port,
+      int shuffleId,
+      MergeFinalizerListener listener) {
+    checkInit();
+    try {
+      TransportClient client = clientFactory.createClient(host, port);
+      ByteBuffer finalizeShuffleMerge = new FinalizeShuffleMerge(appId, shuffleId).toByteBuffer();
+      client.sendRpc(finalizeShuffleMerge, new RpcResponseCallback() {
+        @Override
+        public void onSuccess(ByteBuffer response) {
+          listener.onShuffleMergeSuccess(
+            (MergeStatuses) BlockTransferMessage.Decoder.fromByteBuffer(response));
+        }
+
+        @Override
+        public void onFailure(Throwable e) {
+          listener.onShuffleMergeFailure(e);
+        }
+      });
+    } catch (Exception e) {
+      logger.error("Exception while sending finalizeShuffleMerge request to {}:{}",
```

Comment:
I am uncertain here: is not this one kind of shuffle merge fail too? 
So if you do not call `onShuffleMergeFailure` here then this error is not propagated. 
Would not this cause an endless waiting for some kind of asynchronous result at the caller?

Reply:
Yes, this is a good catch.
`onShuffleMergeFailure` should be called here as well.
It won't cause an endless waiting in the current implementation though, as the caller (a separate thread in DAGScheduler) would bound the wait for getting merge results from all related shuffle service locations.

Reply:
Added the onShuffleMergeFailure.